src/main/java/org/myrobotlab/service/Mqtt.java

```diff
@@ -14,231 +18,390 @@
       log.error("codec failed to initialize", e);
     }
+    // connector = new MqttConnector(this);
   }
 
-  public void connect(String url) {
-    connect(url, null, null);
+  public void autoSubscribe(boolean b) {
+    autoConnect = b;
   }
 
-  synchronized public void connect(String url, String userName, char[] password) {
+  public void broadcastConnect() {
+    try {
+
+      Message onConnect = Message.createMessage(String.format("%s@%s", getName(), getId()), null, "onConnect", new Object[] { Platform.getLocalInstance() });
+      onConnect.sendingMethod = "onConnect";
+      sendRemote(onConnect);
 
-    while (!isConnected) {
+    } catch (MqttException ex) {
       try {
+        if (ex.getReasonCode() == 32104) {
+          log.warn("client is not connected - attempting to reconnect");
+          client.reconnect();
+        } else {
+          log.error("unknown error", ex);
+        }
+      } catch (Exception e) {
+        log.error("reconnect threw", e);
+      }
+    } catch (Exception e) {
+      error(e);
+    }
+  }
 
-        persistence = new MemoryPersistence();
-        client = new MqttAsyncClient(url, clientId, persistence);
+  synchronized public void connect() {
+    try {
+      persistence = new MemoryPersistence();
 
-        this.url = url;
-        this.userName = userName;
-        this.password = password;
+      URI uri = new URI(url);
+      String scheme = uri.getScheme();
+      if (scheme.equalsIgnoreCase("mqtts")) {
+        url = "ssl" + url.substring(5);
+      } else if (scheme.equalsIgnoreCase("mqtt")) {
+        url = "tcp" + url.substring(4);
+      }
+      client = new MqttAsyncClient(url, clientId, persistence);
 
-        if (userName != null) {
-          conOpt.setUserName(userName);
-          conOpt.setPassword(password);
-        }
+      if (username != null && username.length() > 0) {
+        conOpt.setUserName(username);
+        conOpt.setPassword(password.toCharArray());
+      }
 
-        client.setCallback(this);
-        client.connect(conOpt, userContext, this);
-        int i = 0;
+      client.setCallback(this);
 
-        // wait for a CONNACK !
-        while (!client.isConnected() && i < 10) {
-          sleep(100);
-          i += 1;
-        }
+      // onOpt.setMqttVersion(MqttConnectOptions.MQTT_VERSION_3_1);
+      // conOpt.setMqttVersion(MqttConnectOptions.MQTT_VERSION_3_1_1);
 
-        if (client.isConnected()) {
-          isConnected = true;
-        } else {
-          isConnected = false;
-          if (!autoReconnect) {
-            broadcastState();
-            return;
-          }
-        }
+      if (sslCertFilePath != null && sslKeyFilePath != null) {
+        SSLSocketFactory socketFactory = SslUtil.getSocketFactory(sslCaFilePath, sslCertFilePath, sslKeyFilePath, sslPassword);
+        conOpt.setSocketFactory(socketFactory);
+      }
+
+      // default is 10 - which drops messages on a fast local network :(
+      conOpt.setMaxInflight(1000);

```

Comment:
maybe make this configurable?  if it's important.

Reply:
ya

Reply:
I don't think its worth exposing at this point ...